File: es8/es8.go
package main

import (
	"fmt"
	"os"
)

type Node struct {
	Children []*Node
	Metadata []int
}

func (n *Node) findLicense1() int {
	sum := 0
	for _, m := range n.Metadata {
		sum += m
	}

	for _, c := range n.Children {
		sum += c.findLicense1()
	}

	return sum
}

func (n *Node) findLicense2() int {
	sum := 0
	if len(n.Children) == 0 {
		for _, m := range n.Metadata {
			sum += m
		}
	} else {
		for _, m := range n.Metadata {
			if m > 0 && m <= len(n.Children) {
				sum += n.Children[m-1].findLicense2()
			}
		}
	}
	return sum
}

func buildNode(input *os.File) *Node {
	var ccount, mcount int
	_, err := fmt.Fscanf(input, "%d %d", &ccount, &mcount)
	if err != nil {
		panic(err)
	}

	n := Node{
		Children: make([]*Node, ccount),
		Metadata: make([]int, mcount),
	}

	for i := 0; i < ccount; i++ {
		n.Children[i] = buildNode(input)
	}

	for i := 0; i < mcount; i++ {
		_, err = fmt.Fscanf(input, "%d", &n.Metadata[i])
		if err != nil {
			panic(err)
		}
	}

	return &n
}

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <filename>\n", os.Args[0])
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	root := buildNode(input)
	if root != nil {
		fmt.Println("License1:", root.findLicense1())
		fmt.Println("License2:", root.findLicense2())
	}

	return 0
}

File: es8/es8.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
	struct node **children;
	size_t ccount;

	size_t *metadata;
	size_t mcount;
};

static void node_free(struct node *n)
{
	for (size_t i = 0; i < n->ccount; i++)
		node_free(n->children[i]);
	free(n);
}

static struct node *node_new(FILE *input)
{
	size_t ccount, mcount;
	if (fscanf(input, " %zu %zu", &ccount, &mcount) != 2)
		return NULL;

	struct node *n = malloc(
		sizeof(*n) +
		ccount * sizeof(n->children[0]) +
		mcount * sizeof(n->metadata[0]));
	if (!n)
		return NULL;

	n->children = (void *)((char *)n + sizeof(*n));
	n->ccount = ccount;
	n->metadata = (void *)((char *)n + sizeof(*n) + ccount * sizeof(n->children[0]));
	n->mcount = mcount;

	for (size_t i = 0; i < ccount; i++)
		n->children[i] = node_new(input);

	for (size_t i = 0; i < mcount; i++) {
		if (fscanf(input, " %zu", &n->metadata[i]) != 1) {
			node_free(n);
			return NULL;
		}
	}

	return n;
}

static int find_license1(struct node *n)
{
	int license = 0;
	for (size_t i = 0; i < n->ccount; i++)
		license += find_license1(n->children[i]);

	for (size_t i = 0; i < n->mcount; i++)
		license += n->metadata[i];

	return license;
}

static int find_license2(struct node *n)
{
	int license = 0;
	if (!n->ccount) {
		license = 0;
		for (size_t i = 0; i < n->mcount; i++) {
			license += n->metadata[i];
		}
		return license;
	} else {
		license = 0;
		for (size_t i = 0; i < n->mcount; i++) {
			size_t j = n->metadata[i];
			if (j > 0 && j <= n->ccount)
				license += find_license2(n->children[j-1]);
		}
		return license;
	}
}

int main(int argc, char *argv[])
{
	if (argc < 2) {
		fprintf(stderr, "Usage: %s [filename]\n", argv[0]);
		return -1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input) {
		fprintf(stderr, "Cannot open %s for reading\n", argv[1]);
		return -1;
	}

	struct node *root = node_new(input);
	if (root) {
		printf("License1: %d\n", find_license1(root));
		printf("License2: %d\n", find_license2(root));
		node_free(root);
	}

	fclose(input);
	return 0;
}
